packages/flutter/lib/src/material/navigation_bar.dart

```diff
@@ -0,0 +1,1180 @@
+    final ThemeData theme = Theme.of(context);
+    final ColorScheme colorScheme = theme.colorScheme;
+    final NavigationBarThemeData navigationBarTheme = NavigationBarTheme.of(context);
+    final Animation<double> animation = NavigationDestinationInfo.of(context).selectedAnimation;
+
+    return NavigationDestinationBuilder(
+      label: label,
+      tooltip: tooltip,
+      buildIcon: (BuildContext context) {
+        final IconThemeData defaultIconTheme = IconThemeData(
+          size: 24,
+          color: colorScheme.onSurface,
+        );
+        final Widget selectedIconWidget = IconTheme.merge(
+          data: navigationBarTheme.iconTheme?.resolve(<MaterialState>{MaterialState.selected}) ?? defaultIconTheme,
+          child: selectedIcon ?? icon,
+        );
+        final Widget unselectedIconWidget = IconTheme.merge(
+          data: navigationBarTheme.iconTheme?.resolve(<MaterialState>{}) ?? defaultIconTheme,
+          child: icon,
+        );
+
+        return Stack(
+          alignment: Alignment.center,
+          children: <Widget>[
+            NavigationIndicator(
+              animation: animation,
+              color: navigationBarTheme.indicatorColor,
+            ),
+            _StatusTransitionWidgetBuilder(
+              animation: animation,
+              builder: (BuildContext context, Widget? child) {
+                return animation.isForwardOrCompleted
+                    ? selectedIconWidget
+                    : unselectedIconWidget;
+              },
+            ),
+          ],
+        );
+      },
+      buildLabel: (BuildContext context) {
+        final TextStyle? defaultTextStyle = theme.textTheme.overline?.copyWith(
+          color: colorScheme.onSurface,
+        );
+        final TextStyle? effectiveSelectedLabelTextStyle = navigationBarTheme.labelTextStyle?.resolve(<MaterialState>{MaterialState.selected}) ?? defaultTextStyle;
+        final TextStyle? effectiveUnselectedLabelTextStyle = navigationBarTheme.labelTextStyle?.resolve(<MaterialState>{}) ?? defaultTextStyle;
+        return Padding(
+          padding: const EdgeInsets.only(top: 4),
+          child: _ClampTextScaleFactor(
+            // Don't scale labels of destinations, instead, tooltip text will
+            // upscale.
+            upperLimit: 1,
+            child: Text(
+              label,
+              style: animation.isForwardOrCompleted
+                  ? effectiveSelectedLabelTextStyle
+                  : effectiveUnselectedLabelTextStyle,
+            ),
+          ),
+        );
+      },
+    );
+  }
+}
+
+/// Selected Indicator from the Material 3 Navigation Bar component.
+///
+/// When [animation] is 0, the indicator is not present. As [animation] grows
+/// from 0 to 1, the indicator scales in on the x axis.
+///
+/// Useful in a [Stack] widget behind the icons in the Material 3 Navigation Bar
+/// to illuminate the selected destination.
+class NavigationIndicator extends StatelessWidget {
```

Comment:
Made private for now

The generic name is because we would eventually want to use this in the `NavigationRail` too as it is getting a similar redesign